google/cloud/bigquery/job/query.py

```diff
@@ -1300,12 +1300,47 @@ def result(
                 If the job did not complete in the given timeout.
         """
         try:
-            super(QueryJob, self).result(retry=retry, timeout=timeout)
+            retry_do_query = getattr(self, "_retry_do_query", None)
+            first = True
+            sub_retry = retry if retry_do_query is None else None
+
+            def do_get_result():
+                nonlocal first
+
+                if first:
+                    first = False
+                else:
+                    # Note that we won't get here if retry_do_query is
+                    # None, because we won't use a retry.
+
+                    # The orinal job is failed. Create a new one.
+                    job = retry_do_query()
+
+                    # If it's already failed, we might as well stop:
```

Comment:
We don't want to stop after a single retry, right? I'd expect there to be a retry object that inspects the exception to see if it's able to reissue the query.

```suggestion
                    # If it's already failed, we might as well let the query retry handle the failure:
```